### PR TITLE
Use tox to run test and manage test dependencies

### PR DESCRIPTION
**Pull request by doismellburning**
- Explicit test dependencies with version pinning
- Automatic virtualenv management

(Does introduce an implicit tox dependency for contributors!)

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -0,0 +1,11 @@
+[tox]
+skipsdist = True
+envlist =
+	py27,
```
Comment (lurch) [inline tox.ini]: Could / Should you add the rest of the python versions that we support in here?

**Comment by lurch** [comment]
Dunno if I did something wrong, but on my Ubuntu laptop I just tried testing this change by grabbing a local copy of this PR and running `tox`, and it printed a big list of errors, the most relevant of which appears to be `DistributionNotFound: gpiozero`

```
py27 create: /home/andrew/github/RPi-distro/python-gpiozero/.tox/py27
py27 installdeps: -r/home/andrew/github/RPi-distro/python-gpiozero/requirements-test.txt
py27 installed: argparse==1.2.1,coverage==4.2,funcsigs==1.0.2,mock==2.0.0,pbr==1.10.0,py==1.4.31,pytest==3.0.2,six==1.10.0,wsgiref==0.1.2
py27 runtests: PYTHONHASHSEED='4174249863'
py27 runtests: commands[0] | /home/andrew/github/RPi-distro/python-gpiozero/.tox/py27/bin/coverage run -m py.test tests -v
============================= test session starts ==============================
platform linux2 -- Python 2.7.6, pytest-3.0.2, py-1.4.31, pluggy-0.3.1 -- /home/andrew/github/RPi-distro/python-gpiozero/.tox/py27/bin/python2.7
cachedir: .cache
rootdir: /home/andrew/github/RPi-distro/python-gpiozero, inifile: 
collected 0 items / 11 errors 

==================================== ERRORS ====================================
____________________ ERROR collecting tests/test_boards.py _____________________
tests/test_boards.py:14: in <module>
    from gpiozero.pins.mock import MockPin, MockPWMPin
gpiozero/__init__.py:59: in <module>
    from .devices import (
gpiozero/devices.py:63: in <module>
    pin_factory = _default_pin_factory()
gpiozero/devices.py:51: in _default_pin_factory
    dist = pkg_resources.get_distribution('gpiozero')
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:346: in get_distribution
    if isinstance(dist,Requirement): dist = get_provider(dist)
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:225: in get_provider
    return working_set.find(moduleOrReq) or require(str(moduleOrReq))[0]
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:685: in require
    needed = self.resolve(parse_requirements(requirements))
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:588: in resolve
    raise DistributionNotFound(req)
E   DistributionNotFound: gpiozero
____________________ ERROR collecting tests/test_compat.py _____________________
tests/test_compat.py:13: in <module>
    from gpiozero.compat import *
gpiozero/__init__.py:59: in <module>
    from .devices import (
gpiozero/devices.py:63: in <module>
    pin_factory = _default_pin_factory()
gpiozero/devices.py:51: in _default_pin_factory
    dist = pkg_resources.get_distribution('gpiozero')
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:346: in get_distribution
    if isinstance(dist,Requirement): dist = get_provider(dist)
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:225: in get_provider
    return working_set.find(moduleOrReq) or require(str(moduleOrReq))[0]
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:685: in require
    needed = self.resolve(parse_requirements(requirements))
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:588: in resolve
    raise DistributionNotFound(req)
E   DistributionNotFound: gpiozero
____________________ ERROR collecting tests/test_devices.py ____________________
tests/test_devices.py:12: in <module>
    from gpiozero.pins.mock import MockPin
gpiozero/__init__.py:59: in <module>
    from .devices import (
gpiozero/devices.py:63: in <module>
    pin_factory = _default_pin_factory()
gpiozero/devices.py:51: in _default_pin_factory
    dist = pkg_resources.get_distribution('gpiozero')
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:346: in get_distribution
    if isinstance(dist,Requirement): dist = get_provider(dist)
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:225: in get_provider
    return working_set.find(moduleOrReq) or require(str(moduleOrReq))[0]
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:685: in require
    needed = self.resolve(parse_requirements(requirements))
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:588: in resolve
    raise DistributionNotFound(req)
E   DistributionNotFound: gpiozero
____________________ ERROR collecting tests/test_inputs.py _____________________
tests/test_inputs.py:14: in <module>
    from gpiozero.pins.mock import (
gpiozero/__init__.py:59: in <module>
    from .devices import (
gpiozero/devices.py:63: in <module>
    pin_factory = _default_pin_factory()
gpiozero/devices.py:51: in _default_pin_factory
    dist = pkg_resources.get_distribution('gpiozero')
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:346: in get_distribution
    if isinstance(dist,Requirement): dist = get_provider(dist)
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:225: in get_provider
    return working_set.find(moduleOrReq) or require(str(moduleOrReq))[0]
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:685: in require
    needed = self.resolve(parse_requirements(requirements))
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:588: in resolve
    raise DistributionNotFound(req)
E   DistributionNotFound: gpiozero
___________________ ERROR collecting tests/test_mock_pin.py ____________________
tests/test_mock_pin.py:14: in <module>
    from gpiozero.pins.mock import MockPin, MockPWMPin
gpiozero/__init__.py:59: in <module>
    from .devices import (
gpiozero/devices.py:63: in <module>
    pin_factory = _default_pin_factory()
gpiozero/devices.py:51: in _default_pin_factory
    dist = pkg_resources.get_distribution('gpiozero')
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:346: in get_distribution
    if isinstance(dist,Requirement): dist = get_provider(dist)
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:225: in get_provider
    return working_set.find(moduleOrReq) or require(str(moduleOrReq))[0]
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:685: in require
    needed = self.resolve(parse_requirements(requirements))
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:588: in resolve
    raise DistributionNotFound(req)
E   DistributionNotFound: gpiozero
____________________ ERROR collecting tests/test_outputs.py ____________________
tests/test_outputs.py:15: in <module>
    from gpiozero.compat import isclose
gpiozero/__init__.py:59: in <module>
    from .devices import (
gpiozero/devices.py:63: in <module>
    pin_factory = _default_pin_factory()
gpiozero/devices.py:51: in _default_pin_factory
    dist = pkg_resources.get_distribution('gpiozero')
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:346: in get_distribution
    if isinstance(dist,Requirement): dist = get_provider(dist)
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:225: in get_provider
    return working_set.find(moduleOrReq) or require(str(moduleOrReq))[0]
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:685: in require
    needed = self.resolve(parse_requirements(requirements))
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:588: in resolve
    raise DistributionNotFound(req)
E   DistributionNotFound: gpiozero
___________________ ERROR collecting tests/test_pins_data.py ___________________
tests/test_pins_data.py:14: in <module>
    import gpiozero.devices
gpiozero/__init__.py:59: in <module>
    from .devices import (
gpiozero/devices.py:63: in <module>
    pin_factory = _default_pin_factory()
gpiozero/devices.py:51: in _default_pin_factory
    dist = pkg_resources.get_distribution('gpiozero')
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:346: in get_distribution
    if isinstance(dist,Requirement): dist = get_provider(dist)
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:225: in get_provider
    return working_set.find(moduleOrReq) or require(str(moduleOrReq))[0]
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:685: in require
    needed = self.resolve(parse_requirements(requirements))
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:588: in resolve
    raise DistributionNotFound(req)
E   DistributionNotFound: gpiozero
___________________ ERROR collecting tests/test_real_pins.py ___________________
tests/test_real_pins.py:18: in <module>
    from gpiozero import PinFixedPull, PinInvalidPull, PinInvalidFunction
gpiozero/__init__.py:59: in <module>
    from .devices import (
gpiozero/devices.py:63: in <module>
    pin_factory = _default_pin_factory()
gpiozero/devices.py:51: in _default_pin_factory
    dist = pkg_resources.get_distribution('gpiozero')
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:346: in get_distribution
    if isinstance(dist,Requirement): dist = get_provider(dist)
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:225: in get_provider
    return working_set.find(moduleOrReq) or require(str(moduleOrReq))[0]
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:685: in require
    needed = self.resolve(parse_requirements(requirements))
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:588: in resolve
    raise DistributionNotFound(req)
E   DistributionNotFound: gpiozero
______________________ ERROR collecting tests/test_spi.py ______________________
tests/test_spi.py:15: in <module>
    from gpiozero import *
gpiozero/__init__.py:59: in <module>
    from .devices import (
gpiozero/devices.py:63: in <module>
    pin_factory = _default_pin_factory()
gpiozero/devices.py:51: in _default_pin_factory
    dist = pkg_resources.get_distribution('gpiozero')
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:346: in get_distribution
    if isinstance(dist,Requirement): dist = get_provider(dist)
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:225: in get_provider
    return working_set.find(moduleOrReq) or require(str(moduleOrReq))[0]
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:685: in require
    needed = self.resolve(parse_requirements(requirements))
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:588: in resolve
    raise DistributionNotFound(req)
E   DistributionNotFound: gpiozero
__________________ ERROR collecting tests/test_spi_devices.py __________________
tests/test_spi_devices.py:16: in <module>
    from gpiozero.compat import isclose
gpiozero/__init__.py:59: in <module>
    from .devices import (
gpiozero/devices.py:63: in <module>
    pin_factory = _default_pin_factory()
gpiozero/devices.py:51: in _default_pin_factory
    dist = pkg_resources.get_distribution('gpiozero')
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:346: in get_distribution
    if isinstance(dist,Requirement): dist = get_provider(dist)
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:225: in get_provider
    return working_set.find(moduleOrReq) or require(str(moduleOrReq))[0]
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:685: in require
    needed = self.resolve(parse_requirements(requirements))
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:588: in resolve
    raise DistributionNotFound(req)
E   DistributionNotFound: gpiozero
_____________________ ERROR collecting tests/test_tools.py _____________________
tests/test_tools.py:14: in <module>
    from gpiozero.tools import *
gpiozero/__init__.py:59: in <module>
    from .devices import (
gpiozero/devices.py:63: in <module>
    pin_factory = _default_pin_factory()
gpiozero/devices.py:51: in _default_pin_factory
    dist = pkg_resources.get_distribution('gpiozero')
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:346: in get_distribution
    if isinstance(dist,Requirement): dist = get_provider(dist)
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:225: in get_provider
    return working_set.find(moduleOrReq) or require(str(moduleOrReq))[0]
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:685: in require
    needed = self.resolve(parse_requirements(requirements))
.tox/py27/local/lib/python2.7/site-packages/pkg_resources.py:588: in resolve
    raise DistributionNotFound(req)
E   DistributionNotFound: gpiozero
!!!!!!!!!!!!!!!!!!! Interrupted: 11 errors during collection !!!!!!!!!!!!!!!!!!!
=========================== 11 error in 2.81 seconds ===========================
ERROR: InvocationError: '/home/andrew/github/RPi-distro/python-gpiozero/.tox/py27/bin/coverage run -m py.test tests -v'
___________________________________ summary ____________________________________
ERROR:   py27: commands failed
```

**Comment by lurch** [comment]
Looking at the Travis build failure for this PR, do you need to modify https://github.com/RPi-Distro/python-gpiozero/blob/master/setup.py#L52 too?

I'm afraid I dunno how / if Travis and tox are supposed to interact... hmm I just did a search and found http://stackoverflow.com/questions/37016688/how-to-run-tox-with-travis-ci

**Comment by waveform80** [comment]
Tox is definitely something I'd like to see supported but, yes, this has got to work nicely with Travis too. I'll try and find some time to dig into this later this week.

**Comment by lurch** [comment]
As well as https://pypi.python.org/pypi/tox-travis I also found https://pypi.python.org/pypi/panci and https://pypi.python.org/pypi/detox , all of which look interesting.

**Comment by doismellburning** [comment]
> Dunno if I did something wrong

@lurch I accidentally left `skipsdist` in there, my bad, sorry

**Comment by doismellburning** [comment]
> Looking at the Travis build failure for this PR, do you need to modify https://github.com/RPi-Distro/python-gpiozero/blob/master/setup.py#L52 too?

@lurch That appears to be a blank line - what commit were you referring to?

**Comment by lurch** [comment]
> That appears to be a blank line

Ahh, the line-numbers all shifted after a recent commit ;)  The line I was referring to was the `__extra_requires__` 'test' section (in setup.py).

**Comment by lurch** [comment]
The Travis-build is currently failing for Py3.2 - I guess your tox.ini needs some kind of special-case code like https://github.com/RPi-Distro/python-gpiozero/blob/master/.travis.yml has?
And I note that .travis.yml lists pypy3 which your tox.ini doesn't?

**Comment by doismellburning** [comment]
> I note that .travis.yml lists pypy3 which your tox.ini doesn't?

That'll teach me to rely on PyPI classifiers and not the Travis config!

**Comment by doismellburning** [comment]
Bleh this may involve some hackery because of how tox handles dependency lists

**Comment by bennuttall** [comment]
Looks like f05efa993e1b8a2c130e785e352e913f6fdada86 resolves this :+1:

**Pull request closed without merging**